### PR TITLE
Improved Hybrid Deployment Topology

### PR DESCRIPTION
**Pull request by mperniola**

### CODE REVIEW

**Comment by alberto-brigandi** [comment]
@mperniola let me know when this PR is ready

**Comment by mperniola** [comment]
@alberto-brigandi the PR is ready; there is a lack of tests but for the moment I think it may be okay ...

File: src/main/java/it/reply/orchestrator/dto/cmdb/ChronosService.java [hunk 1]
```diff
@@ -47,10 +47,11 @@ public ChronosService(
       @Nullable String region,
       @NonNull String hostname,
       @Nullable String parentServiceId,
-      @NonNull ChronosServiceProperties properties,
-      boolean iamEnabled) {
+      boolean iamEnabled,
+      @Nullable boolean publicIpAssignable,
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/ChronosService.java]:
```suggestion
      boolean publicIpAssignable,
```

File: src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java [hunk 2]
```diff
@@ -39,15 +39,27 @@
   @JsonProperty("images")
   @NonNull
   @NotNull
-  @Builder.Default
+  //@Builder.Default
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java]: Why did you remove it?

File: src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java [hunk 3]
```diff
@@ -85,11 +109,18 @@ public ComputeService(
       @Nullable String parentServiceId,
       @NonNull List<Image> images,
       @NonNull List<Flavor> flavors,
-      boolean iamEnabled) {
+      boolean iamEnabled,
+      @Nullable boolean publicIpAssignable,
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java]:
```suggestion
      boolean publicIpAssignable,
```

File: src/main/java/it/reply/orchestrator/dto/cmdb/MarathonService.java [hunk 4]
```diff
@@ -51,10 +51,11 @@ public MarathonService(
       @Nullable String region,
       @NonNull String hostname,
       @Nullable String parentServiceId,
-      @NonNull MarathonServiceProperties properties,
-      boolean iamEnabled) {
+      boolean iamEnabled,
+      @Nullable boolean publicIpAssignable,
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/MarathonService.java]:
```suggestion
      boolean publicIpAssignable,
```

File: src/main/java/it/reply/orchestrator/dto/cmdb/MesosFrameworkService.java [hunk 5]
```diff
@@ -71,10 +75,11 @@ public MesosFrameworkService(
       @Nullable String region,
       @NonNull String hostname,
       @Nullable String parentServiceId,
-      @NonNull T properties,
-      boolean iamEnabled) {
+      boolean iamEnabled,
+      @Nullable boolean publicIpAssignable,
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/MesosFrameworkService.java]:
```suggestion
      boolean publicIpAssignable,
```

File: src/main/java/it/reply/orchestrator/dto/cmdb/QcgService.java [hunk 6]
```diff
@@ -41,8 +66,9 @@ public QcgService(
       @Nullable String region,
       @NonNull String hostname,
       @Nullable String parentServiceId,
-      boolean iamEnabled) {
+      boolean iamEnabled,
+      @Nullable boolean publicIpAssignable) {
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/QcgService.java]:
```suggestion
      boolean publicIpAssignable) {
```

File: src/main/java/it/reply/orchestrator/service/CmdbServiceImpl.java [hunk 7]
```diff
@@ -272,9 +272,14 @@ public CloudProvider fillCloudProviderInfo(String providerId,
                 .collect(Collectors.toList());
             List<Image> imageList = new ArrayList<>();
             List<Flavor> flavorList = new ArrayList<>();
-            for (Tenant tenant : tenantList) {
+            if (!tenantList.isEmpty()) {
+              //only one element must be here
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/service/CmdbServiceImpl.java]: why?

File: src/main/resources/tosca-definitions/normative-types.yml [hunk 8]
```diff
@@ -335,14 +335,14 @@ node_types:
             attached to specific Compute node instance
           capability: tosca.capabilities.network.Bindable
           relationship: tosca.relationships.network.BindsTo
-          node: tosca.nodes.Compute
+#          node: tosca.nodes.Compute
```
Comment (alberto-brigandi) [inline src/main/resources/tosca-definitions/normative-types.yml]: Why are you changing the normative definitions? are they not aligned with the specification?

**Comment by alberto-brigandi** [comment]
Except for a couple of questions, it looks good to me. Good job

File: src/main/java/it/reply/orchestrator/service/CmdbServiceImpl.java (hunk 7 above)
Comment (mperniola) [reply src/main/java/it/reply/orchestrator/service/CmdbServiceImpl.java]: To avoid exception in the following Tenant tenant = tenantList.get(0);
For removed to avoid break (the devil) at the end of first cycle; i need only one (the first) entry

File: src/main/resources/tosca-definitions/normative-types.yml (hunk 8 above)
Comment (mperniola) [reply src/main/resources/tosca-definitions/normative-types.yml]: Sorry, pasted in from Marica or Miguel... Don't remember reason (changed 14/02) Maybe to avoid some malfunctioning

File: src/main/java/it/reply/orchestrator/dto/cmdb/ChronosService.java (hunk 1 above)
Comment (mperniola) [reply src/main/java/it/reply/orchestrator/dto/cmdb/ChronosService.java]: In truth it is nullable even in the CloudService base class ... I defined it nullable to maintain compatibility with existing services (which do not have that setting)

File: src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java (hunk 2 above)
Comment (mperniola) [reply src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java]: Because the decoration has no meaning if the class does not have @Builder decoration

File: src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java (hunk 3 above)
Comment (mperniola) [reply src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java]: see previous comment

File: src/main/java/it/reply/orchestrator/dto/cmdb/MarathonService.java (hunk 4 above)
Comment (mperniola) [reply src/main/java/it/reply/orchestrator/dto/cmdb/MarathonService.java]: see previous comment

File: src/main/java/it/reply/orchestrator/dto/cmdb/MesosFrameworkService.java (hunk 5 above)
Comment (mperniola) [reply src/main/java/it/reply/orchestrator/dto/cmdb/MesosFrameworkService.java]: see previous comment

File: src/main/java/it/reply/orchestrator/service/CmdbServiceImpl.java (hunk 7 above)
Comment (alberto-brigandi) [reply src/main/java/it/reply/orchestrator/service/CmdbServiceImpl.java]: ok, I see

File: src/main/java/it/reply/orchestrator/dto/cmdb/ChronosService.java (hunk 1 above)
Comment (alberto-brigandi) [reply src/main/java/it/reply/orchestrator/dto/cmdb/ChronosService.java]: then it must be removed also from the base class

File: src/main/java/it/reply/orchestrator/dto/cmdb/CloudService.java [hunk 9]
```diff
@@ -91,6 +91,10 @@
   @JsonProperty("iam_enabled")
   private boolean iamEnabled = true;
 
+  @JsonProperty("public_ip_assignable")
+  @Nullable
+  private boolean publicIpAssignable;
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/CloudService.java]:
```suggestion
  private boolean publicIpAssignable;
```

File: src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java [hunk 10]
```diff
@@ -39,15 +39,27 @@
...
-  @Builder.Default
+  //@Builder.Default
   List<Image> images = new ArrayList<>();
 
   @JsonProperty("flavors")
   @NonNull
   @NotNull
-  @Builder.Default
+  //@Builder.Default
   List<Flavor> flavors = new ArrayList<>();
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java]:
```suggestion
  List<Flavor> flavors = new ArrayList<>();
```

File: src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java [hunk 11]
```diff
@@ -39,15 +39,27 @@
   @JsonProperty("images")
   @NonNull
   @NotNull
-  @Builder.Default
+  //@Builder.Default
   List<Image> images = new ArrayList<>();
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java]:
```suggestion
  List<Image> images = new ArrayList<>();
```

File: src/main/java/it/reply/orchestrator/dto/cmdb/ChronosService.java (hunk 1 above)
Comment (mperniola) [reply src/main/java/it/reply/orchestrator/dto/cmdb/ChronosService.java]: I mean you missed it :D What about backward compatibility?

Comment (mperniola) [reply src/main/java/it/reply/orchestrator/dto/cmdb/ChronosService.java]: you are committing changes without answering this question....

File: src/main/java/it/reply/orchestrator/dto/cmdb/MarathonService.java [hunk 12]
```diff
@@ -63,7 +64,7 @@ public MarathonService(
 
     @NonNull
     @NotNull
-    @Builder.Default
+    //@Builder.Default
     @JsonProperty("load_balancer_ips")
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/MarathonService.java]:
```suggestion
    @JsonProperty("load_balancer_ips")
```

File: src/main/java/it/reply/orchestrator/dto/cmdb/MesosFrameworkServiceProperties.java [hunk 13]
```diff
@@ -47,7 +46,7 @@
 
   @NonNull
   @NotNull
-  @Builder.Default
+  //@Builder.Default
   @JsonProperty("persistent_storage_drivers")
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/cmdb/MesosFrameworkServiceProperties.java]:
```suggestion
  @JsonProperty("persistent_storage_drivers")
```

File: src/main/resources/tosca-definitions/normative-types.yml (hunk 8 above)
Comment (alberto-brigandi) [reply src/main/resources/tosca-definitions/normative-types.yml]: If you don't remember the reason, I'd put it in another PR then

File: src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java (hunk 2 above)
Comment (alberto-brigandi) [reply src/main/java/it/reply/orchestrator/dto/cmdb/ComputeService.java]: ok I removed it for good (no point in leaving dead commented code)

File: src/main/resources/tosca-definitions/normative-types.yml [hunk 14]
```diff
@@ -557,7 +557,7 @@ capability_types:
       network_name:
         type: string
         required: false
-        default: PRIVATE
+#        default: PRIVATE
```
Comment (alberto-brigandi) [inline src/main/resources/tosca-definitions/normative-types.yml]:
```suggestion
```

File: src/main/resources/tosca-definitions/normative-types.yml [hunk 15]
```diff
@@ -335,14 +335,14 @@ node_types:
...
-          node: tosca.nodes.Compute
+#          node: tosca.nodes.Compute
       - link:
           description: >
             Link requirement expresses the relationship between Port and Network
             nodes. It indicates which network this port will connect to.
           capability: tosca.capabilities.network.Linkable
           relationship: tosca.relationships.network.LinksTo
-          node: tosca.nodes.network.Network
+#          node: tosca.nodes.network.Network
```
Comment (alberto-brigandi) [inline src/main/resources/tosca-definitions/normative-types.yml]:
```suggestion
```

File: src/main/resources/tosca-definitions/normative-types.yml (hunk 8 above)
Comment (alberto-brigandi) [inline src/main/resources/tosca-definitions/normative-types.yml]:
```suggestion
```
